lib/scheduler/schedulers_test.go

```diff
@@ -0,0 +1,210 @@
+
+	// variable-arrival-rate
+	{`{"varrival": {"type": "variable-arrival-rate", "startRate": 10, "timeUnit": "30s", "preAllocatedVUs": 20, "maxVUs": 50,
+		"stages": [{"duration": "3m", "target": 30}, {"duration": "5m", "target": 10}]}}`,
+		false, false, func(t *testing.T, cm ConfigMap) {
+			sched := NewVariableArrivalRateConfig("varrival")
+			sched.StartRate = null.IntFrom(10)
+			sched.Stages = []Stage{
+				Stage{Target: null.IntFrom(30), Duration: types.NullDurationFrom(180 * time.Second)},
+				Stage{Target: null.IntFrom(10), Duration: types.NullDurationFrom(300 * time.Second)},
+			}
+			sched.TimeUnit = types.NullDurationFrom(30 * time.Second)
+			sched.PreAllocatedVUs = null.IntFrom(20)
+			sched.MaxVUs = null.IntFrom(50)
+			require.Equal(t, cm, ConfigMap{"varrival": sched})
+			assert.Equal(t, int64(50), cm["varrival"].GetMaxVUs())
+			assert.Equal(t, 510*time.Second, cm["varrival"].GetMaxDuration())
+			assert.Empty(t, cm["varrival"].Validate())
+		}},
+	{`{"varrival": {"type": "variable-arrival-rate", "preAllocatedVUs": 20, "maxVUs": 50, "stages": [{"duration": "5m", "target": 10}]}}`, false, false, nil},
+	{`{"varrival": {"type": "variable-arrival-rate", "preAllocatedVUs": -20, "maxVUs": 50, "stages": [{"duration": "5m", "target": 10}]}}`, false, true, nil},
+	{`{"varrival": {"type": "variable-arrival-rate", "startRate": -1, "preAllocatedVUs": 20, "maxVUs": 50, "stages": [{"duration": "5m", "target": 10}]}}`, false, true, nil},
+	{`{"varrival": {"type": "variable-arrival-rate", "preAllocatedVUs": 20, "stages": [{"duration": "5m", "target": 10}]}}`, false, true, nil},
+	{`{"varrival": {"type": "variable-arrival-rate", "maxVUs": 50, "stages": [{"duration": "5m", "target": 10}]}}`, false, true, nil},
+	{`{"varrival": {"type": "variable-arrival-rate", "preAllocatedVUs": 20, "maxVUs": 50}}`, false, true, nil},
+	{`{"varrival": {"type": "variable-arrival-rate", "preAllocatedVUs": 20, "maxVUs": 50, "stages": []}}`, false, true, nil},
+	{`{"varrival": {"type": "variable-arrival-rate", "preAllocatedVUs": 20, "maxVUs": 50, "stages": [{"duration": "5m", "target": 10}], "timeUnit": "-1s"}}`, false, true, nil},
+	{`{"varrival": {"type": "variable-arrival-rate", "preAllocatedVUs": 30, "maxVUs": 20, "stages": [{"duration": "5m", "target": 10}]}}`, false, true, nil},
+}
+
+func TestConfigMapParsingAndValidation(t *testing.T) {
+	for i, tc := range testCases {
+		tc := tc
+		t.Run(fmt.Sprintf("TestCase#%d", i), func(t *testing.T) {
+			var result ConfigMap
+			err := json.Unmarshal([]byte(tc.rawJSON), &result)
+			if tc.expectParseError {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+
+			validationErrors := result.Validate()
+			if tc.expectValidationError {
+				assert.NotEmpty(t, validationErrors)
+			} else {
+				assert.Empty(t, validationErrors)
+			}
+			if tc.customValidator != nil {
+				tc.customValidator(t, result)
```

Comment:
Maybe call `t.Helper()` in all customValidators so that if/when they exit with error the stacktrace is more helpful

Reply:
Hmm I'm not sure I see any difference when I call `t.Helper()` in the custom validator. Maybe it's because we're using testify?

Reply:
And even if I intentionally add a `t.Errorf("blah")` on line 161 for example, without `t.Helper()` before it I get:
```

--- FAIL: TestConfigMapParsingAndValidation (0.03s)
    --- FAIL: TestConfigMapParsingAndValidation/TestCase#58 (0.00s)
        /go/src/github.com/loadimpact/k6/lib/scheduler/schedulers_test.go:161: blah
```
and with `t.Helper()` I get:
```
--- FAIL: TestConfigMapParsingAndValidation (0.03s)
    --- FAIL: TestConfigMapParsingAndValidation/TestCase#58 (0.00s)
        /go/src/github.com/loadimpact/k6/lib/scheduler/schedulers_test.go:206: blah
```
I'd argue that the first one, without `t.Helper()`, is much more useful :wink:

Reply:
hmm apparently I am mistaken ... because somewhere(in k6) when you call multiple functions I got stacktrace of the error when something failed, but maybe this is testify. `t.Helper()` documentation says it will do what you show, and not what I thought it does  so scrap this